zebedee/client.go

```diff
@@ -0,0 +1,302 @@
+)
+
+// Client is an interface for communication to a zebedee server:
+// https://github.com/ONSdigital/zebedee
+//
+// Get will return a response from the server as a byte array, usually in a json
+// format.
+//
+// GetLanding will return a static dataset landing page struct populated with the
+// data from a given path in zebedee.
+type Client interface {
+	Get(path string) ([]byte, error)
+	GetLanding(path string) (StaticDatasetLandingPage, error)
+	SetAccessToken(token string)
+}
+
+type ZebedeeClient struct {
+	zebedeeURL  string
+	client      *http.Client
+	accessToken string
+}
+
+// StaticDatasetLandingPage is a StaticDatasetLandingPage representation
+type StaticDatasetLandingPage datasetLandingPageStatic.Page
+
+// ErrInvalidZebedeeResponse is returned when zebedee does not respond
+// with a valid status
+type ErrInvalidZebedeeResponse struct {
+	err  error
+	data zebedeeRequestErrorData
+}
+
+type zebedeeRequestErrorData map[string]interface{}
+
+func (e ErrInvalidZebedeeResponse) Error() string {
+	return e.err.Error()
+}
+
+var _ error = ErrInvalidZebedeeResponse{}
+
+// NewClient creates a new Zebedee Client
+func NewClient(url string) ZebedeeClient {
+	return ZebedeeClient{
+		zebedeeURL: url,
+		client:     &http.Client{Timeout: 5 * time.Second},
+	}
+}
+
+func (c ZebedeeClient) Get(path string) ([]byte, error) {
+	return c.get(path)
+}
+
+func (c ZebedeeClient) GetLanding(path string) (StaticDatasetLandingPage, error) {
+	b, err := c.get(path)
+	if err != nil {
+		return StaticDatasetLandingPage{}, err
+	}
+
+	dlp := new(models.DatasetLandingPage)
+	if err = json.Unmarshal(b, &dlp); err != nil {
+		return StaticDatasetLandingPage{}, err
+	}
+
+	related := [][]model.Related{
+		dlp.RelatedDatasets,
+		dlp.RelatedDocuments,
+		dlp.RelatedMethodology,
+		dlp.RelatedMethodologyArticle,
+	}
+
+	//Concurrently resolve any URIs where we need more data from another page
+	var wg sync.WaitGroup
+	sem := make(chan int, 10)
+
+	for _, element := range related {
+		for i, e := range element {
+			sem <- 1
+			wg.Add(1)
+			go func(i int, e model.Related, element []model.Related) {
+				defer func() {
+					<-sem
+					wg.Done()
+				}()
+				element[i].Title = c.getPageTitle(e.URI)
+			}(i, e, element)
+		}
+	}
+	wg.Wait()
+
+	return c.populateStaticLandingPageModel(*dlp)
+}
+
+func (c ZebedeeClient) SetAccessToken(token string) {
+	log.Debug("adding access token to client", log.Data{"token": token})
+	c.accessToken = token
+}
+
+func (c ZebedeeClient) get(path string) ([]byte, error) {
+	req, err := http.NewRequest("GET", c.zebedeeURL+path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(c.accessToken) > 0 {
+		req.Header.Set("X-Florence-Token", c.accessToken)
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrInvalidZebedeeResponse{errors.New("unexpected response from zebedee"), zebedeeRequestErrorData{
```

Comment:
should drain the response body here, something like `io.Copy(ioutil.Discard, resp.Body)`